feat(disclosure): Create and document the disclosure (#529)
twilio-labs/paste: packages/paste-website/src/pages/components/disclosure/index.mdx

    also looking for a contributor to symbolize this component in Sketch. Interested? Reach out!
  </CalloutText>
</Callout>

## Guidelines

### About Disclosure

<p>{props.pageContext.frontmatter.description} It uses the [Paste Heading](/components/heading) component to create consistency across pages and content hierarchy.</p>

### Accessibility

You **must not** nest focusable or actionable elements inside the `DisclosureHeading`. Nested actionable elements are difficult to discover for users of assistive technology as their existence is often not announced.

## Examples

### Default Disclosure

The default variant of a Disclosure is a composition of the [Paste Heading](/components/heading) and a chevron button icon as the "trigger" for the disclosure. This trigger controls the appearance of content that is below it.

The `DisclosureHeading` takes the same props as the [Heading](/components/heading) component so you can control the semantic heading level and it's visual appearance through variants.

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
    Stack,
  }}
>
  {`<Stack orientation="vertical" spacing="space70">
  <Disclosure>

>     <DisclosureHeading as="h2" variant="heading10">

Reviewer: ```suggestion
    <DisclosureHeading as="h2" variant="heading30">
```

---
title: Disclosure
package: '@twilio-paste/disclosure'

> description: Use the Disclosure component to create accessible, hierarchical and collapsible structure to your pages.

Author: I'll add this separately as there's a typo in the suggestion

      headings {
        depth
        value
      }
    }
  }
`;

<ComponentHeader
  name={props.pageContext.frontmatter.title}
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/master/packages/paste-core/components/disclosure"
  storybookUrl="https://twilio-labs.github.io/paste/?path=/story/components-disclosure--default"
  data={props.data.allPasteComponent.edges}
/>

---

<contentwrapper>

<TableOfContents headings={props.data.mdx.headings} />

<content>

<Callout>
  <CalloutTitle>Using alpha components in your product</CalloutTitle>
  <CalloutText>
    Alpha components are ready for you to use in production. Components such as Disclosure are likely to not see any API
    changes. This is only an alpha due to the lack of documentation and a matching Sketch symbol. Feel free to share
    your experience using this component by{' '}
    <Anchor href="https://github.com/twilio-labs/paste/discussions/new">starting a discussion on GitHub</Anchor>. We are
    also looking for a contributor to symbolize this component in Sketch. Interested? Reach out!
  </CalloutText>
</Callout>

## Guidelines

### About Disclosure

> <p>{props.pageContext.frontmatter.description}</p>

Reviewer: ```suggestion
<p>{props.pageContext.frontmatter.description}</p> It uses the [Paste Heading](/components/heading) component to create consistency across pages and content hierarchy.
```

Author: same here adding separately

> It uses the [Paste Heading](/components/heading) component to create consistency across pages and content hierarchy.

Reviewer: ```suggestion
```

### Accessibility

> You **must not** place focusable or actionable elements inside of the `DisclosureHeading`. Nested actionable elements are difficult to discover for users of assistive technology as their existance is often not announced.

Reviewer: ```suggestion
You **must not** nest focusable or actionable elements inside the `DisclosureHeading`. Nested actionable elements are difficult to discover for users of assistive technology as their existence is often not announced.
```

## Examples

### Default Disclosure

> The default variant of a Disclosure is a composition of the [Paste Heading](/components/heading), a chevron icon and a button as the "trigger" for the disclosure. This trigger controls the appearance of content that is below it.

Reviewer: ```suggestion
The default variant of a Disclosure is a composition of the [Paste Heading](/components/heading) and a chevron button icon as the "trigger" for the disclosure. This trigger controls the appearance of content that is below it.
```

> The `DisclosureHeading` takes the same props as the [Heading](/components/heading) component so you can control the semantic heading level and it's visual appearance through heading variants.

Reviewer: ```suggestion
The `DisclosureHeading` takes the same props as the [Heading](/components/heading) component so you can control the semantic heading level and it's visual appearance through variants.
```

> <LivePreview

Reviewer: Can we place all of these examples in a single preview?

  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure>
  <DisclosureHeading as="h2" variant="heading10">

>     Disclosure Heading

Reviewer: ```suggestion
    Between the World and Me by Ta-Nehisi Coates
```

  </DisclosureHeading>
  <DisclosureContent>

>     Content

Reviewer: ```suggestion
    But race is the child of racism, not the father. And the process of naming “the people” has never been a matter of genealogy and physiognomy so much as one of hierarchy. Difference in hue and hair is old. But the belief in the preeminence of hue and hair, the notion that these factors can correctly organize a society and that they signify deeper attributes, which are indelible—this is the new idea at the heart of these new people who have been brought up hopelessly, tragically, deceitfully, to believe that they are white.
```

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure>
  <DisclosureHeading as="h2" variant="heading40">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure>
  <DisclosureHeading as="h2" variant="heading50">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure>
  <DisclosureHeading as="h2" variant="heading60">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

### Contained Disclosure

The `contained` variant of the Disclosure takes the default composition and wraps it in the [Paste Card](/components/card) component. In this instance we do not expose the props used by Card, but all previously mentioned Disclosure props behave in the same way as the default variant.

> <LivePreview

Reviewer: Again, let's merge these examples to a single sheet

  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure variant="contained">
  <DisclosureHeading as="h2" variant="heading50">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure variant="contained">
  <DisclosureHeading as="h2" variant="heading60">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

### Controlling initial state

You can control the intial state of the disclosure to be expanded or collapsed by setting the `visible` prop.

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
  }}
>
  {`<Disclosure visible>
  <DisclosureHeading as="h2" variant="heading30">
    Disclosure Heading
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

## States

> ### Disabled

Reviewer: Should we have a disabled pointer for this one?

Author: Frustratingly because the primitive adds an inline style for `pointer-events: none` it removes the ability to change any pointer based styles for that element.

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
    Text,
  }}
>
  {`<Disclosure>
  <DisclosureHeading as="h2" variant="heading30">
    Disclosure Heading <Text color="colorTextWeak" fontSize="fontSize50" as="span">(3 new items)</Text>
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

<LivePreview
  scope={{
    Disclosure,
    DisclosureHeading,
    DisclosureContent,
    ErrorIcon,
    MediaObject,
    MediaFigure,
    MediaBody,
    Text,
  }}
>
  {`<Disclosure variant="contained" visible>
  <DisclosureHeading as="h2" variant="heading30">
    <MediaObject verticalAlign="center">
      <MediaBody>
        Disclosure Heading <Text color="colorTextWeak" fontSize="fontSize50" as="span">- deployment issue</Text>
      </MediaBody>
      <MediaFigure>
        <ErrorIcon iconColor="colorTextError" decorative={false} title="Error" />
      </MediaFigure>
    </MediaObject>
  </DisclosureHeading>
  <DisclosureContent>
    Content
  </DisclosureContent>
</Disclosure>`}
</LivePreview>

## Anatomy

### Disclosure

#### Contained

> Same Anatomy as [Card](/components/card).

Reviewer: ```suggestion
Contained Disclosure variants share the Paste [Card anatomy](/components/card/#anatomy).
```

### DisclosureHeading

> Same Anatomy as [Heading](/components/heading), plus:

Reviewer: So, we don't have the heading anatomy documented here. File a ticket for it?

Author: Separate Ticket
